In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:
df= pd.read_csv(r"C:\Users\shubh\OneDrive\Documents\Udemy ML course\My Datasets\WBC Dataset.csv", header=0)

In [3]:
df.head()

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  texture_worst  perimeter_worst  area_worst  smoothness_worst  \
0  ...          17.33           184.60      2019.0            0.1622   
1  ...          23.41           158.80      1956.0            0.1238   
2  ...          25.53           152.50      1709.0            0.1444   
3  ...          26.50            98.87       567.7            0.2098   
4  ...          16.67           152.20      1575.0            0.1374   

   compactness_worst  concavity_worst  concave points_worst  symmetry_worst  \
0             0.6656           0.7119                0.2654          0.4601   
1             0.1866           0.2416                0.1860          0.2750   
2             0.4245           0.4504                0.2430          0.3613   
3             0.8663           0.6869                0.2575          0.6638   
4             0.2050           0.4000                0.1625          0.2364   

   fractal_dimension_worst  Unnamed: 32  
0                  0.11890          NaN  
1                  0.08902          NaN  
2                  0.08758          NaN  
3                  0.17300          NaN  
4                  0.07678          NaN  

[5 rows x 33 columns]

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

In [5]:
del df["Unnamed: 32"]
del df["id"]

In [6]:
df = pd.get_dummies(df, drop_first=True)

In [7]:
df.head()

radius_mean  texture_mean  perimeter_mean  area_mean  smoothness_mean  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   fractal_dimension_mean  ...  texture_worst  perimeter_worst  area_worst  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  diagnosis_M  
0          0.4601                  0.11890            1  
1          0.2750                  0.08902            1  
2          0.3613                  0.08758            1  
3          0.6638                  0.17300            1  
4          0.2364                  0.07678            1  

[5 rows x 31 columns]

In [8]:
X = df.loc[:, df.columns!='diagnosis_M']
y = df['diagnosis_M']

In [60]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3,random_state= 100)

### Multivariate analysis

In [61]:
from sklearn.feature_selection import f_regression as fr
result = fr(X,y)
f_score = result[0]
p_values = result[1]

# Getting column names 
columns = list(X.columns)
print(" ")
print(" ")
print(" ")

print("     Features                     ","F-Score  ","P-Values")
print("     ------------                   --------  ---------")

for i in range(0,len(columns)):
    f1 = "%4.2f" % f_score[i]
    p1 = "%2.6f" % p_values[i]
    print("    ",columns[i].ljust(25),f1.rjust(12),"",p1.rjust(8))

 
 
 
     Features                      F-Score   P-Values
     ------------                   --------  ---------
     radius_mean                     646.98  0.000000
     texture_mean                    118.10  0.000000
     perimeter_mean                  697.24  0.000000
     area_mean                       573.06  0.000000
     smoothness_mean                  83.65  0.000000
     compactness_mean                313.23  0.000000
     concavity_mean                  533.79  0.000000
     concave points_mean             861.68  0.000000
     symmetry_mean                    69.53  0.000000
     fractal_dimension_mean            0.09  0.759937
     radius_se                       268.84  0.000000
     texture_se                        0.04  0.843332
     perimeter_se                    253.90  0.000000
     area_se                         243.65  0.000000
     smoothness_se                     2.56  0.110297
     compactness_se                   53.25  0.000000
     concavity_se   

### Standardizing the data

In [62]:
from sklearn.preprocessing import StandardScaler

In [63]:
sc = StandardScaler().fit(X_train)

In [64]:
X_train_std = sc.transform(X_train)

In [65]:
X_test_std = sc.transform(X_test)

### Training SVM Classfier (Linear Kernel)

In [90]:
from sklearn import svm

In [91]:
clf_svm_l = svm.SVC(kernel='linear', C=0.01)
clf_svm_l.fit(X_train_std, y_train)

SVC(C=0.01, kernel='linear')

In [92]:
y_train_pred = clf_svm_l.predict(X_train_std)

In [93]:
y_test_pred = clf_svm_l.predict(X_test_std)

In [94]:
from sklearn.metrics import accuracy_score, confusion_matrix

In [95]:
confusion_matrix(y_train, y_train_pred)

array([[254,   1],
       [  9, 134]], dtype=int64)

In [96]:
accuracy_score(y_train, y_train_pred)

0.9748743718592965

In [97]:
confusion_matrix(y_test,  y_test_pred)

array([[102,   0],
       [  7,  62]], dtype=int64)

In [98]:
accuracy_score(y_test, y_test_pred)

0.9590643274853801

In [99]:
clf_svm_l.n_support_

array([45, 45])

### Grid Search

In [100]:
 from sklearn.model_selection import GridSearchCV

In [101]:
params = {
    "C":[0.001,0.005,0.01,0.05,0.1,0.5,1,5,10,50,100,500,1000,5000]
}

In [102]:
clf_svm_l = svm.SVC(kernel="linear")

In [103]:
svm_grid_lin = GridSearchCV(clf_svm_l, params, n_jobs=-1, cv=10, verbose=1, scoring='accuracy')

In [104]:
svm_grid_lin.fit(X_train_std, y_train)

Fitting 10 folds for each of 14 candidates, totalling 140 fits


GridSearchCV(cv=10, estimator=SVC(kernel='linear'), n_jobs=-1,
             param_grid={'C': [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10, 50,
                               100, 500, 1000, 5000]},
             scoring='accuracy', verbose=1)

In [105]:
svm_grid_lin.best_params_

{'C': 0.5}

In [106]:
linsvm_clf = svm_grid_lin.best_estimator_

In [107]:
accuracy_score(y_test, linsvm_clf.predict(X_test_std))

0.9707602339181286

Accuracy Score After Grid Search = 97.076%

### SVM Classifer (Polynomial Kernel)

In [113]:
clf_svm_p = svm.SVC(kernel="poly", degree=2, C=0.1)
clf_svm_p.fit(X_train_std , y_train)

SVC(C=0.1, degree=2, kernel='poly')

In [114]:
y_train_pred = clf_svm_p.predict(X_train_std)
y_test_pred = clf_svm_p.predict(X_test_std)

In [115]:
accuracy_score(y_train, y_train_pred)

0.7412060301507538

In [116]:
accuracy_score(y_test
              , y_test_pred)

0.7368421052631579

In [117]:
clf_svm_p.n_support_

array([133, 125])

In [119]:
params = {
    "degree":[1,2,3,4,5,6],
    "C":[0.001,0.005,0.01,0.05,0.1,0.5,1,5,10,50,100,500,1000,5000]
}

In [121]:
svm_grid_pol = GridSearchCV(clf_svm_p, params, n_jobs=-1, cv=10, verbose=1, scoring='accuracy')

In [122]:
svm_grid_pol.fit(X_train_std, y_train)

Fitting 10 folds for each of 84 candidates, totalling 840 fits


GridSearchCV(cv=10, estimator=SVC(C=0.1, degree=2, kernel='poly'), n_jobs=-1,
             param_grid={'C': [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 5, 10, 50,
                               100, 500, 1000, 5000],
                         'degree': [1, 2, 3, 4, 5, 6]},
             scoring='accuracy', verbose=1)

In [123]:
svm_grid_pol.best_params_

{'C': 5, 'degree': 1}

In [124]:
g_svm_pol = svm_grid_pol.best_estimator_

In [125]:
accuracy_score(y_test, g_svm_pol.predict(X_test_std))

0.9707602339181286